File: store/src/main/java/fr/jayblanc/mbyte/store/files/entity/Node.java
/*
 * Copyright (C) 2025 Jerome Blanchard <jayblanc@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package fr.jayblanc.mbyte.store.files.entity;

import fr.jayblanc.mbyte.store.files.FileService;
import jakarta.persistence.*;

import java.io.Serializable;
import java.util.Comparator;
import java.util.Objects;

@Entity
@NamedQueries({
        @NamedQuery(name = "Node.findAllChildren", query = "SELECT n FROM Node n WHERE n.parent = :parent"),
        @NamedQuery(name = "Node.findAll", query = "SELECT n FROM Node n"),
        @NamedQuery(name = "Node.findChildrenForName", query = "SELECT n FROM Node n WHERE n.parent = :parent AND n.name = :name"),
        @NamedQuery(name = "Node.countChildren", query = "SELECT count(n) FROM Node n WHERE n.parent = :parent"),
})
@Table(indexes = {
        @Index(name = "parent", columnList = "parent"),
        @Index(name = "parent_name", columnList = "parent, name"),
})
public class Node implements Comparable<Node>, Serializable {

    @Enumerated(EnumType.STRING)
    private Type type;
    @Id
    @Column(length = 50)
    private String id;
    @Column(length = 50)
    private String parent;
    @Version
    private long version;
    private String name;
    @Column(length = 50)
    private String mimetype;
    private long size;
    private long creation;
    private long modification;
    private String content;

    public Node() {
        this.creation = this.modification = System.currentTimeMillis();
        this.size = 0;
    }

    public Node(Type type, String parent, String id, String name) {
        this();
        this.type = type;
        this.parent = parent;
        this.id = id;
        this.name = name;
    }

    public Type getType() {
        return type;
    }

    public void setType(Type type) {
        this.type = type;
    }

    public String getParent() {
        return parent;
    }

    public void setParent(String parent) {
        this.parent = parent;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public long getVersion() {
        return version;
    }

    public void setVersion(long version) {
        this.version = version;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getMimetype() {
        return mimetype;
    }

    public void setMimetype(String mimetype) {
        this.mimetype = mimetype;
    }

    public long getSize() {
        return size;
    }

    public void setSize(long size) {
        this.size = size;
    }

    public long getCreation() {
        return creation;
    }

    public void setCreation(long creation) {
        this.creation = creation;
    }

    public long getModification() {
        return modification;
    }

    public void setModification(long modification) {
        this.modification = modification;
    }

    public boolean isRoot() {
        return this.id.equals(FileService.ROOT_NODE_ID);
    }

    public boolean isFolder() {
        return this.type.equals(Type.TREE);
    }

    public enum Type {
        TREE,
        BLOB
    }

    @Override
    public int compareTo(Node o) {
        if (this.getType().equals(o.getType())) {
            return this.getName().compareTo(o.getName());
        } else if (this.getType().equals(Type.TREE)){
            return 1;
        } else {
            return -1;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Node node = (Node) o;
        return version == node.version && size == node.size && creation == node.creation && modification == node.modification && type == node.type && Objects.equals(id, node.id) && Objects.equals(parent, node.parent) && Objects.equals(name, node.name) && Objects.equals(mimetype, node.mimetype) && Objects.equals(content, node.content);
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, id, parent, version, name, mimetype, size, creation, modification, content);
    }

    @Override
    public String toString() {
        return "Node{" +
                "type=" + type +
                ", id='" + id + '\'' +
                ", version=" + version +
                ", parent=" + parent +
                ", name='" + name + '\'' +
                ", mimetype='" + mimetype + '\'' +
                ", size=" + size +
                ", creation=" + creation +
                ", modification=" + modification +
                ", content='" + content + '\'' +
                '}';
    }

    public static class NameComparatorAsc implements Comparator<Node> {
        @Override
        public int compare(Node o1, Node o2) {
            if ( o1.isFolder() && !o2.isFolder() ) {
                return -1;
            }
            if ( !o1.isFolder() && o2.isFolder() ) {
                return 1;
            }
            return o1.getName().compareTo(o2.getName());
        }
    }

    public static class NameComparatorDesc implements Comparator<Node> {
        @Override
        public int compare(Node o1, Node o2) {
            if ( o1.isFolder() && !o2.isFolder() ) {
                return -1;
            }
            if ( !o1.isFolder() && o2.isFolder() ) {
                return 1;
            }
            return o2.getName().compareTo(o1.getName());
        }
    }
}

File: gui/src/javascript/api/entities/Node.ts
///
/// Copyright (C) 2025 Jerome Blanchard <jayblanc@gmail.com>
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program.  If not, see <https://www.gnu.org/licenses/>.
///

// Frontend representation of a store Node (renamed to Node)
// Mirrors the NodeDto shape and adds convenience helpers.

// Internal DTO type (NOT exported) to avoid exposing NodeDto separately
type NodeData = {
  root?: boolean
  type?: string
  id: string
  parent?: string | null
  name?: string
  mimetype?: string | null
  size?: number
  creation?: string | null
  modification?: string | null
}

export default class Node {
  readonly dto: NodeData

  constructor(dto: NodeData) {
    this.dto = dto
  }

  static fromDto(dto: NodeData): Node {
    return new Node(dto)
  }

  get id(): string {
    return this.dto.id
  }

  get name(): string {
    return this.dto.name ?? ''
  }

  get mimetype(): string | undefined {
    return this.dto.mimetype ?? undefined
  }

  get size(): number | undefined {
    return this.dto.size
  }

  get creationTs(): number | undefined {
    if (!this.dto.creation) return undefined
    const t = Date.parse(this.dto.creation)
    return Number.isNaN(t) ? undefined : t
  }

  get modificationTs(): number | undefined {
    if (!this.dto.modification) return undefined
    const t = Date.parse(this.dto.modification)
    return Number.isNaN(t) ? undefined : t
  }

  get isRoot(): boolean {
    return Boolean(this.dto.root)
  }

  get isFolder(): boolean {
    return (this.dto.type ?? '').toUpperCase() === 'TREE'
  }

  get isFile(): boolean {
    return (this.dto.type ?? '').toUpperCase() === 'BLOB'
  }
}
